### ISSUE TITLE
IndentationError: expected an indented block

### ISSUE THREAD

**Issue by snuggles42** [iago-lito/intim#11]
This simple script causes the error in the subject (like every loop I try to execute) (python3, xterm):

for t in range(10):
    print(str(t))
          ^

There are no hidden tabs etc. in the script and the script runs without problems when directly processed with python. Any ideas?

**Comment by iago-lito** [comment]
This is not an Intim error, but a python error. In python, you need to indent each nested block or you get an `IndentationError`. In other words, the following should work:

    for t in range(10):
        print(str(t))
    ^^^^ indentation

Please refer to [python documentation](https://docs.python.org/3/) for information about this scripting language. Semantic whitespace is one of its particularities.

Once the indentation is correct. You may select the whole loop in vim visual mode (for instance with `vip`) then press `<space>` to send all lines at once with `<Plug>IntimSendChunk`. The indentation should be sent along correctly (see `:help intim-passing` for more information).

**Comment by iago-lito** [comment]
If your interpreter is `sage` or `ipython`, you may want to read about the `MagicCpaste` feature in `:help intim-autoindent`.

Also, in order to initialize the loop (but not run it completely), `git pull` to the latest release bf19f37 and read `:help intim-loop` :)

**Comment by snuggles42** [comment]
I know about indentation in Python. Your example as "test.py" with "python3 test.py" works perfectly. But when using intim I still get the indentation error. Any idea?

>>> for t in range(10):

... print(str(t))

  File "<stdin>", line 2

    print(str(t))

        ^

IndentationError: expected an indented block

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by iago-lito** [comment]
Sorry, I must have badly explained myself :\ Let me put this another way :)

**TL;DR:** skip to the end.

---

Because of semantic indentation in python, Intim *removes* the indentation before any line you send with `<Plug>IntimSendLine`. The reason for this is that, if your cursor were here: 

    for i in range(10):
        print(i)
        ^

Then you wouldn't be able to send `>>> print(i)` with the spacebar, because `>>>     print(i)` would be sent instead, which would result in an `IndentationError`.

In order not to experience this, Intim removes any indentation before a line you send as a single line, in order to avoid the `IndentationError`. 
The consequence is that, if your cursor stands here:

    for i in range(10):
    ^   print(i)

and you press `<space><space>`, the lines sent are actually `>>> for i in range(10):` (ok) and `... print(i)` (not ok: indent error because python expects `...     print(i)` at this point). Thus the error you are experiencing.

So, if you want to send the whole loop, you have to do it another way. For example, as I wrote earlier:

> You may select the whole loop in vim visual mode (for instance with `vip`) then press `<space>` to send all lines at once with `<Plug>IntimSendChunk`. The indentation should be sent along correctly (see `:help intim-passing` for more information).

The solution is therefore to use `<Plug>IntimSendChunk` instead of `<Plug>IntimSendLine` or `<Plug>IntimSendSelection` in this case. By default, it is mapped to visual `<space>`.

When you do this, Intim does not send your lines directly. Instead, it saves them to a temporary `chunk` file, then asks python to read this file. While doing so, it does __not__ remove indentation. So it should work as you expect.

---

In a nutshell, in this vim situation, and with the default mappings:

```
    for i in range(4):
    ^   print(i)
```

press `vj$<space>` and you should see

```
0
1
2
3
```

printed to your python console.

**Comment by snuggles42** [comment]
Perfect. Works like a charm. Thanks so much for your detailed help. I think I will even replace N-vimR when working with R. A big leap forward!

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
